onnx/defs/nn/defs.cc

```diff
@@ -1432,4 +1433,144 @@ ONNX_OPERATOR_SET_SCHEMA(
+        .Attr(
+            "ngram_indexes",
+            "list of int64s (type: AttributeProto::INTS). This list is parallel to the specified 'pool_*' attribute. "
+            "The i-th element in ngram_indexes indicate the coordinate of the i-th n-gram in the output tensor.",
+            AttributeProto::INTS)
+        .Attr(
+            "weights",
+            "list of floats. This attribute stores the weight of each n-gram in pool. The i-th element in weights "
+            "is the weight of the i-th n-gram in pool. Its length equals to the size of ngram_indexes. "
+            "By default, weights is an all-one tensor.This attribute is used when mode is \"IDF\" or \"TFIDF\" "
+            "to scale the associated word counts.",
+            AttributeProto::FLOATS,
+            OPTIONAL)
+        .Attr(
+            "mode",
+            "The weighting criteria. It can be one of \"TF\" (term frequency), "
+            "\"IDF\" (inverse document frequency), and \"TFIDF\" (the combination of TF and IDF)",
+            AttributeProto::STRING)
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          auto output_elem_type = ctx.getOutputType(0)->mutable_tensor_type();
+          output_elem_type->set_elem_type(TensorProto::FLOAT);
+
+          if (hasInputShape(ctx, 0)) {
+            std::vector<int64_t> ngram_indexes;
+            getRepeatedAttribute(ctx, "ngram_indexes", ngram_indexes);
+            if (ngram_indexes.empty() ||
+                !std::all_of(
+                    ngram_indexes.cbegin(),
+                    ngram_indexes.cend(),
+                    [](int64_t i) { return i >= 0; })) {
+              fail_shape_inference(
+                  "ngram_indexes must be non-empty with no negative values");
+            }
+
+            auto greatest_hit =
+                std::max_element(ngram_indexes.cbegin(), ngram_indexes.cend());
+            auto max_last_axis = *greatest_hit + 1;
+
+            TensorShapeProto output_shape;
+            auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
+            auto dim_size = input_shape.dim_size();
+            if (dim_size == 0 || dim_size == 1) {
+              output_shape.add_dim()->set_dim_value(max_last_axis);
+            } else if (dim_size == 2) {
+              auto& B_dim = input_shape.dim(0);
+              if (B_dim.has_dim_value()) {
+                output_shape.add_dim()->set_dim_value(B_dim.dim_value());
+              } else if (B_dim.has_dim_param()) {
+                output_shape.add_dim()->set_dim_param(B_dim.dim_param());
+              } else {
+                fail_shape_inference(
+                    "Input shape does not have first dimension value");
+              }
+              output_shape.add_dim()->set_dim_value(max_last_axis);
+            } else {
+              fail_shape_inference(
+                  "Input shape must have either [C], [?], [B,C] or [?,C] dimensions where C > 0 and B > 0");
```

Comment:
May be "Input tensor must have rank 1 or 2" or "Input tensor must have rank <= 2" depending on whether rank 0 (scalar) is permitted.